rust_src/src/data.rs

```diff
@@ -789,4 +790,36 @@ pub fn logcount(value: EmacsInt) -> i32 {
     unsafe { rust_count_one_bits(value as usize) }
 }
 
+/// Parse STRING as a decimal number and return the number.
+/// Ignore leading spaces and tabs, and all trailing chars.  Return 0 if
+/// STRING cannot be parsed as an integer or floating point number.
+///
+/// If BASE, interpret STRING as a number in that base.  If BASE isn't
+/// present, base 10 is used.  BASE must be between 2 and 16 (inclusive).
+/// If the base used is not 10, STRING is always parsed as an integer.
+#[lisp_fn(min = "1", name = "string-to-number", c_name = "string_to_number")]
+pub fn string_to_number_lisp(mut string: LispStringRef, base: Option<EmacsInt>) -> LispObject {
+    let b = match base {
+        None => 10,
+        Some(n) => {
+            if !(n >= 2 && n <= 16) {
+                xsignal!(Qargs_out_of_range, base.into());
+            }
+            n
+        }
+    };
+
+    let p = string.sdata_ptr();
+    unsafe {
+        while *p == ' ' as c_char || *p == '\t' as c_char {
+            *p += 1;
+        }
+    }
+
```

Comment:
You can replace the the string_to_number call with this to return an EmacsInt.

```
+    unsafe { string_to_number(p, b as i32, true) }
+        .as_fixnum()
+        .unwrap_or(0)
```

Reply:
So we cast the returned LispObject from string_to_number to fixnum and let lisp_fn cast it back to LispObject ? You once told me we only use explicit return types when the types allow it.

Reply:
When I looked deeper into the C, `string_to_number` can actually return a Float or an Int. So this function should really return `LispNumber`.

As we are moving forward with the ports, we are trying to have all of the functions return native Rust types where possible which allows for the Rust code to perform computations without constantly flipping from LispObject -> Rust -> LispObject.